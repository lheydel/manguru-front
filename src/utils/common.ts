import { AxiosRequestConfig } from 'axios';

export const requestConfig: AxiosRequestConfig = {
    withCredentials: true,
    headers: {
        'Content-Type': 'application/json'
    }
};
