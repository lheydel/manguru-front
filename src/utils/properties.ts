export enum Language {
    DEFAULT = 'en',
    EN = 'en',
    FR = 'fr'
}
