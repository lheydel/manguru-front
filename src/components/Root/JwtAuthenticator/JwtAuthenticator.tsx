import React from 'react';
import { Cookies, withCookies } from 'react-cookie';
import { UserLoginState } from '../../../services/user/reducers/user.states';
import { connect } from 'react-redux';
import { ApplicationState } from '../../../services/common/app.states';
import userActionners from '../../../services/user/actions/user.actionners';
import { Cookie } from '../../../utils/properties';

export interface JwtAuthenticatorProps {
    cookies: Cookies;
    loginJwt: typeof userActionners.loginJwtRequest;
    loginState: UserLoginState;
}

/**
 * Ensure that a user always stays authenticated if a JWT is set in its cookies
 */
export const RawJwtAuthenticator: React.FC<JwtAuthenticatorProps> = (props) => {
    const { cookies, loginJwt } = props;
    const { logged, error, loading } = props.loginState;

    // if user not logged but has an Authorization cookie, try to authenticate him
    if (!logged && !error && cookies.get(Cookie.AUTH)) {
        if (!loading) {
            loginJwt();
        }
        return null;
    } // else

    return <React.Fragment>{props.children}</React.Fragment>;
};

const mapStateToProps = (state: ApplicationState) => ({
    loginState: state.user.login
});

const mapDispatchToProps = (dispatch: Function) => ({
    loginJwt: () => dispatch(userActionners.loginJwtRequest()),
});

export const JwtAuthenticator = withCookies(connect(mapStateToProps, mapDispatchToProps)(RawJwtAuthenticator));
