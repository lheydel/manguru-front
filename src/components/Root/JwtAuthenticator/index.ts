import { JwtAuthenticator } from './JwtAuthenticator';
export default JwtAuthenticator;
export * from './JwtAuthenticator';
