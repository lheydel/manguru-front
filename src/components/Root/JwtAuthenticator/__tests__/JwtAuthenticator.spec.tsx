import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Cookies, CookiesProvider } from 'react-cookie';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import createMockStore from 'redux-mock-store';
import JwtAuthenticator from '..';
import { ApplicationState, initialAppState } from '../../../../services/common/app.state';
import { Cookie } from '../../../../utils/properties';
import { RawJwtAuthenticator } from '../JwtAuthenticator';

describe('JwtAuthenticator component', () => {
    let wrapper: ReturnType<typeof connectedWrapper>;

    const authState: (logged: boolean, loading: boolean, error?: string) => ApplicationState = (logged, loading, error) => ({
        ...initialAppState,
        user: {
            login: {
                logged: logged,
                loading: loading,
                error: error,
            }
        }
    });

    const connectedWrapper = (store: Store, cookies: Cookies) => {
        return mount(
            <Provider store={store}>
                <CookiesProvider cookies={cookies}>
                    <JwtAuthenticator>
                        <button />
                    </JwtAuthenticator>
                </CookiesProvider>
            </Provider>
        );
    };

    const doTest = (store: Store, cookies: Cookies, callLogin: boolean) => {
        wrapper = connectedWrapper(store, cookies);
        expect(store.dispatch).toHaveBeenCalledTimes(callLogin ? 1 : 0);
        expect(toJson(wrapper.find(RawJwtAuthenticator).render())).toMatchSnapshot();
    };

    afterEach(() => {
        wrapper.unmount();
    });

    test('user logged --> render children', () => {
        // setup store
        const store = createMockStore()(authState(true, false));
        store.dispatch = jest.fn();

        // actual tests
        doTest(store, new Cookies(), false);
    });

    test('user not logged + jwt cookie --> try to authenticate + render null', () => {
        // setup store
        const store = createMockStore()(authState(false, false));
        store.dispatch = jest.fn();

        // setup jwt cookie
        const cookies = new Cookies();
        cookies.set(Cookie.AUTH, 'token');

        // actual tests
        doTest(store, cookies, true);
    });

    test('user not logged + jwt cookie + loading --> render null', () => {
        // setup store
        const store = createMockStore()(authState(false, true));
        store.dispatch = jest.fn();

        // setup jwt cookie
        const cookies = new Cookies();
        cookies.set(Cookie.AUTH, 'token');

        // actual tests
        doTest(store, cookies, false);
    });

    test('user not logged + auth failed --> render children', () => {
        // setup store
        const store = createMockStore()(authState(false, false, 'err'));
        store.dispatch = jest.fn();

        // setup jwt cookie
        const cookies = new Cookies();
        cookies.set(Cookie.AUTH, 'token');

        // actual tests
        doTest(store, cookies, false);
    });

    test('user not logged + no jwt cookie --> render children', () => {
        // setup store
        const store = createMockStore()(authState(false, false));
        store.dispatch = jest.fn();

        // actual tests
        doTest(store, new Cookies(), false);
    });

});
