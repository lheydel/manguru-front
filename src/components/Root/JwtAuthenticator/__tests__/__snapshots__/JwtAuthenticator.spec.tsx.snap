// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JwtAuthenticator component user logged --> render children 1`] = `<button />`;

exports[`JwtAuthenticator component user not logged + auth failed --> render children 1`] = `<button />`;

exports[`JwtAuthenticator component user not logged + jwt cookie + loading --> render null 1`] = `null`;

exports[`JwtAuthenticator component user not logged + jwt cookie --> try to authenticate + render null 1`] = `null`;

exports[`JwtAuthenticator component user not logged + no jwt cookie --> render children 1`] = `<button />`;
